src/components/AnnotationShareControl.tsx

```diff
@@ -0,0 +1,184 @@
+import {
+  CopyIcon,
+  IconButton,
+  Input,
+  InputGroup,
+  Popover,
+  ShareIcon,
+} from '@hypothesis/frontend-shared';
+import { useCallback, useEffect, useRef, useState } from 'preact/hooks';
+
+import type { Group } from '../helpers';
+import type { Annotation } from '../helpers/annotation-metadata';
+import { isPrivate } from '../helpers/permissions';
+import { isIOS } from '../utils/user-agent';
+
+export type AnnotationShareControlProps = {
+  /** The annotation in question */
+  annotation: Annotation;
+  /** Group to which the annotation belongs */
+  group: Group | null;
+
+  /**
+   * Invoked when the URI is copied to the clipboard.
+   * It indicates if copying to the clipboard was successful or not.
+   */
+  onCopy?: (result: { successful: boolean }) => void;
+};
+
+function selectionOverflowsInputElement() {
+  // On iOS the selection overflows the input element
+  // See: https://github.com/hypothesis/client/pull/2799
+  return isIOS();
+}
+
+/**
+ * Retrieve an appropriate sharing link for this annotation.
+ *
+ * If the annotation is on a shareable document (i.e. its document is
+ * web-accessible), prefer the `incontext` (bouncer) link, but fallback to the
+ * `html` (single-annotation `h` web view) link if needed.
+ *
+ * If the annotation is not on a shareable document, don't use the `incontext`
+ * link as that won't work; only use the single-annotation-view `html` link.
+ *
+ * Note that `html` links are not provided by the service for third-party
+ * annotations.
+ */
+function annotationSharingLink(
+  annotation: Annotation,
+  isShareableURI: boolean,
+): string | null {
+  if (isShareableURI) {
+    return annotation.links?.incontext ?? annotation.links?.html ?? null;
+  } else {
+    return annotation.links?.html ?? null;
+  }
+}
+
+/**
+ * Share button which opens a Popover for sharing a single annotation.
+ */
+export default function AnnotationShareControl({
+  annotation,
+  group,
+  onCopy,
+}: AnnotationShareControlProps) {
+  const annotationIsPrivate = isPrivate(annotation.permissions);
+  const inContextAvailable = /^http(s?):/i.test(annotation.uri);
+  const shareURI = annotationSharingLink(annotation, inContextAvailable);
+
+  const inputRef = useRef<HTMLInputElement | null>(null);
+  const shareRef = useRef<HTMLDivElement | null>(null);
+
+  const [isOpen, setOpen] = useState(false);
+  const wasOpen = useRef(isOpen);
+
+  const copyShareLink = useCallback(async () => {
+    try {
+      await navigator.clipboard.writeText(shareURI!);
+      onCopy?.({ successful: true });
+    } catch {
+      onCopy?.({ successful: false });
```

Comment:
The error is thrown away here, I think it would be useful to preserve it for debugging.